lib/faraday/request.rb

```diff
@@ -24,13 +24,14 @@ module Faraday
   #   @return [String] body
   # @!attribute options
   #   @return [RequestOptions] options
-  #
-  # rubocop:disable Style/StructInheritance
-  class Request < Struct.new(:http_method, :path, :params, :headers, :body, :options)
-    # rubocop:enable Style/StructInheritance
-
+  Request = Struct.new(:http_method, :path, :params, :headers, :body, :options) do
     extend MiddlewareRegistry
 
+    alias_method :member_get, :[]
+    private :member_get
+    alias_method :member_set, :[]=
+    private :member_set
```

Comment:
Maybe, I understood why `member_set` is required. Previously, `Request` inherited an anonymous class generated by `Struct`, but in this PR, it becomes just a `Struct` and cannot use `super`. Is it correct?

As far as I see, `#params=` and `#headers=` don't follow the documentation: " Replace params, preserving the existing hash type." In other words, their implementations are different from their documents. I think they don't have to be explicitly declared by overriding themselves. What about simply removing `#params=` and `#headers=` and delegating the functionalities to `Struct`? `Struct` has "#attribute=" out of the box.

Reply:
You are correct for why I added `member_set` :)

I believe the docs for `#params=` and `#headers=` are correct. Normally you don't build a `Request` directly but use methods on `Faraday::Connection`. `Connection` builds `Utils::ParamsHash` and `Utils::Headers` (both subclasses of Hash) in the initializer and [passes](https://github.com/lostisland/faraday/blob/2477fe2da3477559819ccee3ec8aeedf5301d0f9/lib/faraday/connection.rb#L454-L456) these to `Faraday::Request.create`. If you merge in additional params or headers (probably using a regular Hash) like this:
```ruby
connection.get do |req|
  request.headers = { foo: "bar" }
end
```
this code ensures `headers` stays a `Utils::Headers` instance (and does things like stringifying the key) instead of becoming a regular Hash with symbol keys.

Removing these overridden setters leads to one (not very clear) test failure on this line: https://github.com/lostisland/faraday/blob/2477fe2da3477559819ccee3ec8aeedf5301d0f9/spec/faraday/connection_spec.rb#L744
```
  1) Faraday::Connection request params with url and extra params allows to override all params
     Failure/Error: expect(req.params['b']).to eq('b')
     
       expected: "b"
            got: nil
     
       (compared using ==)
```

Reply:
Thank you both for adding even more context to this detail!

Reply:
Thank you @bdewater, and sorry if this struct is a mess 🙄.
Can't wait to get rid of them in v3.0 🙌!

Reply:
Thank you! That's a great explanation 👍 The document seems to be good because it says "preserving the existing hash **type**".

So, `member_set ` is aliased as Struct's original `#[]=` before it is overridden.

https://github.com/lostisland/faraday/blob/2477fe2da3477559819ccee3ec8aeedf5301d0f9/lib/faraday/request.rb#L96-L98

LGTM